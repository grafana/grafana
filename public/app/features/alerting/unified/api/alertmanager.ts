import { lastValueFrom } from 'rxjs';

import { isObject } from '@grafana/data';
import { getBackendSrv, isFetchError } from '@grafana/runtime';
import {
  AlertManagerCortexConfig,
  AlertmanagerGroup,
  AlertmanagerStatus,
  Receiver,
  TestReceiversAlert,
  TestReceiversPayload,
  TestReceiversResult,
} from 'app/plugins/datasource/alertmanager/types';

import { GRAFANA_RULES_SOURCE_NAME, getDatasourceAPIUid } from '../utils/datasource';

// "grafana" for grafana-managed, otherwise a datasource name
export async function fetchAlertManagerConfig(alertManagerSourceName: string): Promise<AlertManagerCortexConfig> {
  try {
    const result = await lastValueFrom(
      getBackendSrv().fetch<AlertManagerCortexConfig>({
        url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/alerts`,
        showErrorAlert: false,
        showSuccessAlert: false,
      })
    );
    return {
      template_files: result.data.template_files ?? {},
      template_file_provenances: result.data.template_file_provenances ?? {},
      alertmanager_config: result.data.alertmanager_config ?? {},
      last_applied: result.data.last_applied,
      id: result.data.id,
    };
  } catch (e) {
    // if no config has been uploaded to grafana, it returns error instead of latest config
    if (
      alertManagerSourceName === GRAFANA_RULES_SOURCE_NAME &&
      isFetchError(e) &&
      e.data?.message?.includes('could not find an Alertmanager configuration')
    ) {
      return {
        template_files: {},
        alertmanager_config: {},
      };
    }
    throw e;
  }
}

export async function updateAlertManagerConfig(
  alertManagerSourceName: string,
  config: AlertManagerCortexConfig
): Promise<void> {
  await lastValueFrom(
    getBackendSrv().fetch({
      method: 'POST',
      url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/alerts`,
      data: config,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );
}

export async function deleteAlertManagerConfig(alertManagerSourceName: string): Promise<void> {
  await lastValueFrom(
    getBackendSrv().fetch({
      method: 'DELETE',
      url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/alerts`,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );
}

export async function fetchAlertGroups(alertmanagerSourceName: string): Promise<AlertmanagerGroup[]> {
  const result = await lastValueFrom(
    getBackendSrv().fetch<AlertmanagerGroup[]>({
      url: `/api/alertmanager/${getDatasourceAPIUid(alertmanagerSourceName)}/api/v2/alerts/groups`,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );

  return result.data;
}

export async function fetchStatus(alertManagerSourceName: string): Promise<AlertmanagerStatus> {
  const result = await lastValueFrom(
    getBackendSrv().fetch<AlertmanagerStatus>({
      url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/api/v2/status`,
      showErrorAlert: false,
      showSuccessAlert: false,
    })
  );

  return result.data;
}

export async function testReceivers(
  alertManagerSourceName: string,
  receivers: Receiver[],
  alert?: TestReceiversAlert
): Promise<void> {
  const data: TestReceiversPayload = {
    receivers,
    alert,
  };
  try {
    const result = await lastValueFrom(
      getBackendSrv().fetch<TestReceiversResult>({
        method: 'POST',
        data,
        url: `/api/alertmanager/${getDatasourceAPIUid(alertManagerSourceName)}/config/api/v1/receivers/test`,
        showErrorAlert: false,
        showSuccessAlert: false,
      })
    );

    if (receiversResponseContainsErrors(result.data)) {
      throw new Error(getReceiverResultError(result.data));
    }
  } catch (error) {
    if (isFetchError(error) && isTestReceiversResult(error.data) && receiversResponseContainsErrors(error.data)) {
      throw new Error(getReceiverResultError(error.data));
    }

    throw error;
  }
}

function receiversResponseContainsErrors(result: TestReceiversResult) {
  return result.receivers.some((receiver) =>
    receiver.grafana_managed_receiver_configs.some((config) => config.status === 'failed')
  );
}

function isTestReceiversResult(data: unknown): data is TestReceiversResult {
  if (isObject(data) && 'receivers' in data && Array.isArray(data.receivers)) {
    return data.receivers.every(isSingleTestRecieverResult);
  }

  return false;
}

function isSingleTestRecieverResult(receiver: unknown): receiver is TestReceiversResult {
  return (
    isObject(receiver) &&
    'name' in receiver &&
    typeof receiver.name === 'string' &&
    'grafana_managed_receiver_configs' in receiver &&
    Array.isArray(receiver.grafana_managed_receiver_configs)
  );
}

function getReceiverResultError(receiversResult: TestReceiversResult) {
  return receiversResult.receivers
    .flatMap((receiver) =>
      receiver.grafana_managed_receiver_configs
        .filter((receiver) => receiver.status === 'failed')
        .map((receiver) => receiver.error ?? 'Unknown error.')
    )
    .join('; ');
}
