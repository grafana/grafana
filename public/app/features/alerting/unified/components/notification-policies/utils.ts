import { ControllerRenderProps } from 'react-hook-form';

import { SelectableValue } from '@grafana/data';
import { ContactPointWithMetadata } from 'app/features/alerting/unified/components/contact-points/utils';

export const handleContactPointSelect = (
  value: SelectableValue<ContactPointWithMetadata>,
  onChange: ControllerRenderProps['onChange']
) => {
  if (value === null) {
    return onChange(null);
  }

  if (!value) {
    return onChange('');
  }

  return onChange(value.value?.name);
};
