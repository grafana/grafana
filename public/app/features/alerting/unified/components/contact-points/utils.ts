import { difference, groupBy, take, trim, upperFirst } from 'lodash';
import { ReactNode } from 'react';

import { config } from '@grafana/runtime';
import { contextSrv } from 'app/core/core';
import { canAdminEntity, shouldUseK8sApi } from 'app/features/alerting/unified/utils/k8s/utils';
import {
  AlertManagerCortexConfig,
  GrafanaManagedContactPoint,
  GrafanaManagedReceiverConfig,
  MatcherOperator,
  Receiver,
  Route,
} from 'app/plugins/datasource/alertmanager/types';
import { NotifierDTO, NotifierStatus, ReceiversStateDTO } from 'app/types';

import { OnCallIntegrationDTO } from '../../api/onCallApi';
import { computeInheritedTree } from '../../utils/notification-policies';
import { extractReceivers } from '../../utils/receivers';
import { ReceiverTypes } from '../receivers/grafanaAppReceivers/onCall/onCall';
import { getOnCallMetadata, ReceiverPluginMetadata } from '../receivers/grafanaAppReceivers/useReceiversMetadata';

import { RECEIVER_META_KEY, RECEIVER_PLUGIN_META_KEY, RECEIVER_STATUS_KEY } from './constants';

const AUTOGENERATED_RECEIVER_POLICY_MATCHER_KEY = '__grafana_receiver__';

// TODO we should really add some type information to these receiver settings...
export function getReceiverDescription(receiver: ReceiverConfigWithMetadata): ReactNode | undefined {
  if (!receiver.settings) {
    return undefined;
  }
  const { settings } = receiver;
  switch (receiver.type) {
    case 'email': {
      const addresses = settings.addresses || settings.to; // when dealing with alertmanager email_configs we don't normalize the settings
      return addresses ? summarizeEmailAddresses(addresses) : undefined;
    }
    case 'slack': {
      const recipient = settings.recipient || settings.channel;
      if (!recipient) {
        return;
      }

      // Slack channel name might have a "#" in the recipient already
      const channelName = recipient.replace(/^#/, '');
      return `#${channelName}`;
    }
    case 'kafka': {
      return settings.kafkaTopic;
    }
    case 'webhook': {
      return settings.url;
    }
    case ReceiverTypes.OnCall: {
      return receiver[RECEIVER_PLUGIN_META_KEY]?.description;
    }
    default:
      return receiver[RECEIVER_META_KEY]?.description;
  }
}

// input: foo+1@bar.com, foo+2@bar.com, foo+3@bar.com, foo+4@bar.com
// output: foo+1@bar.com, foo+2@bar.com, +2 more
export function summarizeEmailAddresses(addresses: string): string {
  const MAX_ADDRESSES_SHOWN = 3;
  const SUPPORTED_SEPARATORS = /,|;|\n+/g;

  // split all email addresses
  const emails = addresses.trim().split(SUPPORTED_SEPARATORS).map(trim);

  // grab the first 3 and the rest
  const summary = take(emails, MAX_ADDRESSES_SHOWN);
  const rest = difference(emails, summary);

  if (rest.length) {
    summary.push(`+${rest.length} more`);
  }

  return summary.join(', ');
}

// Grafana Managed contact points have receivers with additional diagnostics
export interface ReceiverConfigWithMetadata extends GrafanaManagedReceiverConfig {
  // we're using a symbol here so we'll never have a conflict on keys for a receiver
  // we also specify that the diagnostics might be "undefined" for vanilla Alertmanager
  [RECEIVER_STATUS_KEY]?: NotifierStatus | undefined;
  [RECEIVER_META_KEY]: {
    name: string;
    description?: string;
  };
  // optional metadata that comes from a particular plugin (like Grafana OnCall)
  [RECEIVER_PLUGIN_META_KEY]?: ReceiverPluginMetadata;
}

export interface ContactPointWithMetadata extends GrafanaManagedContactPoint {
  id: string;
  policies?: RouteReference[]; // now is optional as we don't have the data from the read-only endpoint
  grafana_managed_receiver_configs: ReceiverConfigWithMetadata[];
}

type EnhanceContactPointsArgs = {
  status?: ReceiversStateDTO[];
  notifiers?: NotifierDTO[];
  onCallIntegrations?: OnCallIntegrationDTO[] | undefined | null;
  contactPoints: Receiver[];
  alertmanagerConfiguration?: AlertManagerCortexConfig;
};

/**
 * This function adds the status information for each of the integrations (contact point types) in a contact point
 * 1. we iterate over all contact points
 * 2. for each contact point we "enhance" it with the status or "undefined" for vanilla Alertmanager
 * contactPoints: list of contact points
 * alertmanagerConfiguration: optional as is passed when we need to get number of policies for each contact point
 * and we prefer using the data from the read-only endpoint.
 */
export function enhanceContactPointsWithMetadata({
  status = [],
  notifiers = [],
  onCallIntegrations,
  contactPoints,
  alertmanagerConfiguration,
}: EnhanceContactPointsArgs): ContactPointWithMetadata[] {
  // compute the entire inherited tree before finding what notification policies are using a particular contact point
  const fullyInheritedTree = computeInheritedTree(alertmanagerConfiguration?.alertmanager_config?.route ?? {});
  const usedContactPoints = getUsedContactPoints(fullyInheritedTree);
  const usedContactPointsByName = groupBy(usedContactPoints, 'receiver');

  const enhanced = contactPoints.map((contactPoint) => {
    const receivers = extractReceivers(contactPoint);
    const statusForReceiver = status.find((status) => status.name === contactPoint.name);

    const id = 'id' in contactPoint && contactPoint.id ? contactPoint.id : contactPoint.name;

    return {
      ...contactPoint,
      id,
      policies:
        alertmanagerConfiguration && usedContactPointsByName && (usedContactPointsByName[contactPoint.name] ?? []),
      grafana_managed_receiver_configs: receivers.map((receiver, index) => {
        const isOnCallReceiver = receiver.type === ReceiverTypes.OnCall;
        // if we don't have alertmanagerConfiguration we can't get the metadata for oncall receivers,
        // because we don't have the url, as we are not using the alertmanager configuration
        // but the contact points returned by the read only permissions contact points endpoint (/api/v1/notifications/receivers)
        return {
          ...receiver,
          [RECEIVER_STATUS_KEY]: statusForReceiver?.integrations[index],
          [RECEIVER_META_KEY]: getNotifierMetadata(notifiers, receiver),
          // if OnCall plugin is installed, we'll add it to the receiver's plugin metadata
          [RECEIVER_PLUGIN_META_KEY]: isOnCallReceiver
            ? getOnCallMetadata(onCallIntegrations, receiver, Boolean(alertmanagerConfiguration))
            : undefined,
        };
      }),
    };
  });

  return enhanced.sort((a, b) => a.name.localeCompare(b.name));
}

export function isAutoGeneratedPolicy(route: Route) {
  const simplifiedRoutingToggleEnabled = config.featureToggles.alertingSimplifiedRouting ?? false;
  if (!simplifiedRoutingToggleEnabled) {
    return false;
  }
  if (!route.object_matchers) {
    return false;
  }
  return (
    route.object_matchers.some((objectMatcher) => {
      return (
        objectMatcher[0] === AUTOGENERATED_RECEIVER_POLICY_MATCHER_KEY && objectMatcher[1] === MatcherOperator.equal
      );
    }) ?? false
  );
}

export interface RouteReference {
  receiver: string;
  route: {
    type: 'auto-generated' | 'normal';
  };
}

export function getUsedContactPoints(route: Route): RouteReference[] {
  const childrenContactPoints = route.routes?.flatMap((route) => getUsedContactPoints(route)) ?? [];

  if (route.receiver) {
    return [
      {
        receiver: route.receiver,
        route: {
          type: isAutoGeneratedPolicy(route) ? 'auto-generated' : 'normal',
        },
      },
      ...childrenContactPoints,
    ];
  }

  return childrenContactPoints;
}

function getNotifierMetadata(notifiers: NotifierDTO[], receiver: GrafanaManagedReceiverConfig) {
  const match = notifiers.find((notifier) => notifier.type === receiver.type);

  return {
    name: match?.name ?? upperFirst(receiver.type),
    description: match?.description,
  };
}

export const showManageContactPointPermissions = (alertmanager: string, contactPoint: GrafanaManagedContactPoint) =>
  shouldUseK8sApi(alertmanager) && contextSrv.licensedAccessControlEnabled() && canAdminEntity(contactPoint);
