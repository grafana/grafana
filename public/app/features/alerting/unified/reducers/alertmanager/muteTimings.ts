import { createAction, createReducer } from '@reduxjs/toolkit';
import { remove } from 'lodash';

import { AlertManagerCortexConfig, MuteTimeInterval } from 'app/plugins/datasource/alertmanager/types';

import { removeTimeIntervalFromRoute, renameTimeInterval } from '../../utils/alertmanager';

export const addMuteTimingAction = createAction<{ interval: MuteTimeInterval }>('muteTiming/add');
export const updateMuteTimingAction = createAction<{
  interval: MuteTimeInterval;
  originalName: string;
}>('muteTiming/update');
export const deleteMuteTimingAction = createAction<{ name: string }>('muteTiming/delete');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to mute timings and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const muteTimingsReducer = createReducer(initialState, (builder) => {
  builder
    // add a mute timing to the alertmanager configuration
    .addCase(addMuteTimingAction, (draft, { payload }) => {
      const { interval } = payload;
      draft.alertmanager_config.time_intervals = (draft.alertmanager_config.time_intervals ?? []).concat(interval);
    })
    // add a mute timing to the alertmanager configuration
    // make sure we update the mute timing in either the deprecated or the new time intervals property
    .addCase(updateMuteTimingAction, (draft, { payload }) => {
      const { interval, originalName } = payload;
      const nameHasChanged = interval.name !== originalName;

      const timeIntervals = draft.alertmanager_config.time_intervals ?? [];
      const muteTimeIntervals = draft.alertmanager_config.mute_time_intervals ?? [];

      const existingIntervalIndex = timeIntervals.findIndex(({ name }) => name === originalName);
      if (existingIntervalIndex !== -1) {
        timeIntervals[existingIntervalIndex] = interval;
      }

      const existingMuteIntervalIndex = muteTimeIntervals.findIndex(({ name }) => name === originalName);
      if (existingMuteIntervalIndex !== -1) {
        muteTimeIntervals[existingMuteIntervalIndex] = interval;
      }

      if (nameHasChanged && draft.alertmanager_config.route) {
        draft.alertmanager_config.route = renameTimeInterval(
          interval.name,
          originalName,
          draft.alertmanager_config.route
        );
      }
    })
    // delete a mute timing from the alertmanager configuration, since the configuration might be using the "deprecated" mute_time_intervals
    // let's also check there
    .addCase(deleteMuteTimingAction, (draft, { payload }) => {
      const { name } = payload;
      const { alertmanager_config } = draft;
      const { time_intervals = [], mute_time_intervals = [] } = alertmanager_config;

      // remove the mute timings from the legacy and new time intervals definition
      remove(time_intervals, (interval) => interval.name === name);
      remove(mute_time_intervals, (interval) => interval.name === name);

      // remove the mute timing from all routes
      alertmanager_config.route = removeTimeIntervalFromRoute(name, alertmanager_config.route ?? {});
    });
});
