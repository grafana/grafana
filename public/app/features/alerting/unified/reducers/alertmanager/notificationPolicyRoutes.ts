import { createAction, createReducer } from '@reduxjs/toolkit';

import { AlertManagerCortexConfig } from 'app/plugins/datasource/alertmanager/types';

import { FormAmRoute } from '../../types/amroutes';
import { addUniqueIdentifierToRoute } from '../../utils/amroutes';
import {
  InsertPosition,
  addRouteToReferenceRoute,
  cleanRouteIDs,
  mergePartialAmRouteWithRouteTree,
  omitRouteFromRouteTree,
} from '../../utils/routeTree';

export const updateRouteAction = createAction<{ update: Partial<FormAmRoute>; alertmanager: string }>('routes/update');
export const deleteRouteAction = createAction<{ id: string }>('routes/delete');
export const addRouteAction = createAction<{
  alertmanager: string;
  partialRoute: Partial<FormAmRoute>;
  referenceRouteIdentifier: string;
  insertPosition: InsertPosition;
}>('routes/add');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to routes and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const routesReducer = createReducer(initialState, (builder) => {
  builder
    // update an existing route
    .addCase(updateRouteAction, (draft, { payload }) => {
      const { update, alertmanager } = payload;
      const { alertmanager_config } = draft;

      if (!alertmanager_config.route) {
        return;
      }

      const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(alertmanager_config.route);
      const newRouteTree = mergePartialAmRouteWithRouteTree(alertmanager, update, rootRouteWithIdentifiers);
      alertmanager_config.route = cleanRouteIDs(newRouteTree);
    })
    // delete a route
    .addCase(deleteRouteAction, (draft, { payload }) => {
      const { id } = payload;
      const { alertmanager_config } = draft;

      // if we don't even have a root route, we bail
      if (!alertmanager_config.route) {
        return;
      }

      const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(alertmanager_config.route);
      const updatedPolicyTree = omitRouteFromRouteTree(id, rootRouteWithIdentifiers);
      draft.alertmanager_config.route = cleanRouteIDs(updatedPolicyTree);
    })
    // add a new route to given position
    .addCase(addRouteAction, (draft, { payload }) => {
      const { partialRoute, referenceRouteIdentifier, insertPosition, alertmanager } = payload;
      const { alertmanager_config } = draft;

      if (!alertmanager_config.route) {
        return;
      }

      const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(alertmanager_config.route);
      const updatedPolicyTree = addRouteToReferenceRoute(
        alertmanager,
        partialRoute,
        referenceRouteIdentifier,
        rootRouteWithIdentifiers,
        insertPosition
      );

      draft.alertmanager_config.route = cleanRouteIDs(updatedPolicyTree);
    });
});
