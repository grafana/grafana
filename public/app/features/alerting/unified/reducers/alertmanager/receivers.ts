import { createAction, createReducer } from '@reduxjs/toolkit';
import { remove } from 'lodash';

import { AlertManagerCortexConfig, Receiver } from 'app/plugins/datasource/alertmanager/types';

import { renameReceiverInRoute } from '../../utils/notification-policies';

export const addReceiverAction = createAction<Receiver>('receiver/add');
export const updateReceiverAction = createAction<{ name: string; receiver: Receiver }>('receiver/update');
export const deleteReceiverAction = createAction<string>('receiver/delete');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to receiver (Contact points) and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const receiversReducer = createReducer(initialState, (builder) => {
  builder
    // add a new receiver
    .addCase(addReceiverAction, (draft, { payload: newReceiver }) => {
      // ensure the receivers are always an array
      const currentReceivers = (draft.alertmanager_config.receivers = draft.alertmanager_config.receivers ?? []);

      // check if the name doesn't already exist
      const nameExists = currentReceivers.some((receiver) => receiver.name === newReceiver.name);
      if (nameExists) {
        throw new Error(`Duplicate receiver name ${newReceiver.name}`);
      }

      // add the receiver
      currentReceivers.push(newReceiver);
    })
    // upate an existing receiver
    .addCase(updateReceiverAction, (draft, { payload }) => {
      const { name, receiver } = payload;
      const renaming = name !== receiver.name;

      const receivers = draft.alertmanager_config.receivers ?? [];

      const targetIndex = receivers.findIndex((receiver) => receiver.name === name);
      const targetExists = targetIndex > -1;

      // check if the receiver we want to update exists
      if (!targetExists) {
        throw new Error(`Expected receiver ${name} to exist, but did not find it in the config`);
      }

      // check if the new name doesn't already exist
      if (renaming) {
        const nameExists = receivers.some((oldReceiver) => oldReceiver.name === receiver.name);
        if (nameExists) {
          throw new Error(`Duplicate receiver name ${receiver.name}`);
        }
      }

      // overwrite the receiver with the new one
      receivers[targetIndex] = receiver;

      // check if we need to update routes if the contact point was renamed
      const routeTree = draft.alertmanager_config.route;

      if (routeTree && renaming) {
        draft.alertmanager_config.route = renameReceiverInRoute(routeTree, name, receiver.name);
      }
    })
    // delete a receiver from the alertmanager configuration
    .addCase(deleteReceiverAction, (draft, { payload: name }) => {
      remove(draft.alertmanager_config.receivers ?? [], (receiver) => receiver.name === name);
    });
});
