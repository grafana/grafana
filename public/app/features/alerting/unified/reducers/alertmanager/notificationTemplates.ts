import { createAction, createReducer } from '@reduxjs/toolkit';
import { remove, toArray, unset } from 'lodash';

import { AlertManagerCortexConfig } from 'app/plugins/datasource/alertmanager/types';

import { TemplateFormValues } from '../../components/receivers/TemplateForm';
import { ensureDefine } from '../../utils/templates';

export const addNotificationTemplateAction = createAction<{ template: TemplateFormValues }>('notificationTemplate/add');
export const updateNotificationTemplateAction = createAction<{
  name: string;
  template: TemplateFormValues;
}>('notificationTemplate/update');
export const deleteNotificationTemplateAction = createAction<{ name: string }>('notificationTemplate/delete');

const initialState: AlertManagerCortexConfig = {
  alertmanager_config: {},
  template_files: {},
};

/**
 * This reducer will manage action related to notification templates and make sure all operations on the alertmanager
 * configuration happen immutably and only mutate what they need.
 */
export const notificationTemplatesReducer = createReducer(initialState, (builder) => {
  builder
    .addCase(addNotificationTemplateAction, (draft, { payload }) => {
      const { alertmanager_config = {}, template_files = {} } = draft;
      const { template } = payload;

      const targetTemplateExists = template_files[template.title] !== undefined;
      if (targetTemplateExists) {
        throw new Error('target template already exists');
      }

      // wrap content in "define" if it's not already wrapped, in case user did not do it/
      // it's not obvious that this is needed for template to work
      const content = ensureDefine(template.title, template.content);

      // add the template to the list of template files
      template_files[template.title] = content;

      // add the template to the alertmanager_config
      alertmanager_config.templates = toArray(alertmanager_config.templates).concat(template.title);
    })
    .addCase(updateNotificationTemplateAction, (draft, { payload }) => {
      const { alertmanager_config = {}, template_files = {} } = draft;
      const { name, template } = payload;
      const renaming = name !== template.title;

      const targetExists = template_files[name] !== undefined;
      if (!targetExists) {
        throw new Error(`Expected notification template ${name} to exist, but did not find it in the config`);
      }

      // wrap content in "define" if it's not already wrapped, in case user did not do it/
      // it's not obvious that this is needed for template to work
      const content = ensureDefine(template.title, template.content);

      if (renaming) {
        const oldName = name;
        const newName = template.title;

        const targetExists = template_files[newName] !== undefined;
        if (targetExists) {
          throw new Error(`Duplicate template name ${newName}`);
        }

        unset(template_files, oldName);
        remove(alertmanager_config.templates ?? [], (templateName) => templateName === oldName);
        alertmanager_config.templates = toArray(alertmanager_config.templates).concat(template.title);
      }

      template_files[template.title] = content;
    })
    .addCase(deleteNotificationTemplateAction, (draft, { payload }) => {
      const { name } = payload;
      const { alertmanager_config = {}, template_files = {} } = draft;

      unset(template_files, name);
      remove(alertmanager_config.templates ?? [], (templateName) => templateName === name);
    });
});
