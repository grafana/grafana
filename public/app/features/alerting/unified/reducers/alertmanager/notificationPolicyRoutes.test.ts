import { produce } from 'immer';

import { AlertManagerCortexConfig, MatcherOperator, Route } from 'app/plugins/datasource/alertmanager/types';

import { FormAmRoute } from '../../types/amroutes';
import { addUniqueIdentifierToRoute } from '../../utils/amroutes';

import { addRouteAction, deleteRouteAction, routesReducer, updateRouteAction } from './notificationPolicyRoutes';

describe('routes', () => {
  const defaultRoute: Route = {
    receiver: 'ROOT',
    group_by: ['grafana_folder'],
    object_matchers: [],

    routes: [
      {
        receiver: 'A',

        object_matchers: [['team', MatcherOperator.equal, 'operations']],
        routes: [
          {
            receiver: 'B1',
            object_matchers: [['region', MatcherOperator.equal, 'europe']],
          },
          {
            receiver: 'B2',
            object_matchers: [['region', MatcherOperator.equal, 'nasa']],
          },
        ],
      },
      {
        receiver: 'C',
        object_matchers: [['foo', MatcherOperator.equal, 'bar']],
      },
    ],
    group_wait: '10s',
    group_interval: '1m',
  };
  const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(defaultRoute);
  const ROOT_identifier = rootRouteWithIdentifiers.id;
  const A_identifier = rootRouteWithIdentifiers.routes![0].id;

  const initialConfig: AlertManagerCortexConfig = {
    alertmanager_config: {
      time_intervals: [],
      mute_time_intervals: [],
      route: defaultRoute,
    },
    template_files: {},
  };
  it('Should add a new route with receiver E as a child of default route', () => {
    const routeWithID = addUniqueIdentifierToRoute({
      receiver: 'E',
    });
    const newFormRoute: Partial<FormAmRoute> = {
      receiver: 'E',
      id: routeWithID.id,
    };
    const action = addRouteAction({
      alertmanager: 'alertmanager',
      partialRoute: newFormRoute,
      referenceRouteIdentifier: ROOT_identifier,
      insertPosition: 'child',
    });

    const result = routesReducer(initialConfig, action);

    expect(result.alertmanager_config.route?.routes).toHaveLength(
      (initialConfig.alertmanager_config.route?.routes?.length ?? 0) + 1
    );

    // the last route if the root route should have receiver E
    expect(result.alertmanager_config.route?.routes?.at(-1)).toMatchObject({
      receiver: 'E',
    });

    // assert the rest of the configuration
    expect(result).toMatchSnapshot();
  });

  it('Should not add a new route with receiver E as a child of route with receiver A, if data has been updated by another user', () => {
    const routeWithID = addUniqueIdentifierToRoute({
      receiver: 'E',
    });
    const newFormRoute: Partial<FormAmRoute> = {
      receiver: 'E',
      id: routeWithID.id,
    };
    const action = addRouteAction({
      alertmanager: 'alertmanager',
      partialRoute: newFormRoute,
      referenceRouteIdentifier: A_identifier,
      insertPosition: 'child',
    });
    const modifiedConfig = produce(initialConfig, (draft) => {
      draft.alertmanager_config.route!.routes = [];
    });
    const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(modifiedConfig.alertmanager_config.route!);
    const modifiedConfigWithIdentifiers = produce(modifiedConfig, (draft) => {
      draft.alertmanager_config.route = rootRouteWithIdentifiers;
    });
    expect(() => routesReducer(modifiedConfigWithIdentifiers, action)).toThrow();
  });

  it('Should update an existing route A with receiver B', () => {
    const newRoute: Partial<FormAmRoute> = {
      receiver: 'B',
      id: A_identifier,
    };
    const action = updateRouteAction({
      alertmanager: 'alertmanager',
      update: newRoute,
    });

    expect(routesReducer(initialConfig, action)).toMatchSnapshot();
  });

  it('Should not update route if config has been updated by another user', () => {
    const newRoute: Partial<FormAmRoute> = {
      receiver: 'B',
      id: A_identifier,
    };
    const action = updateRouteAction({
      alertmanager: 'alertmanager',
      update: newRoute,
    });
    const modifiedConfig = produce(initialConfig, (draft) => {
      draft.alertmanager_config.route!.routes = [];
    });
    const rootRouteWithIdentifiers = addUniqueIdentifierToRoute(modifiedConfig.alertmanager_config.route!);
    const modifiedConfigWithIdentifiers = produce(modifiedConfig, (draft) => {
      draft.alertmanager_config.route = rootRouteWithIdentifiers;
    });
    expect(() => routesReducer(modifiedConfigWithIdentifiers, action)).toThrow();
  });

  it('Should delete route', () => {
    const action = deleteRouteAction({
      id: A_identifier,
    });
    expect(routesReducer(initialConfig, action)).toMatchSnapshot();
  });
});
