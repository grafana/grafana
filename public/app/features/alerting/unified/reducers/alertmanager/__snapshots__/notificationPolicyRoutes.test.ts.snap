// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes Should add a new route with receiver E as a child of default route 1`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [],
    "route": {
      "group_by": [
        "grafana_folder",
      ],
      "group_interval": "1m",
      "group_wait": "10s",
      "object_matchers": [],
      "receiver": "ROOT",
      "routes": [
        {
          "object_matchers": [
            [
              "team",
              "=",
              "operations",
            ],
          ],
          "receiver": "A",
          "routes": [
            {
              "object_matchers": [
                [
                  "region",
                  "=",
                  "europe",
                ],
              ],
              "receiver": "B1",
              "routes": [],
            },
            {
              "object_matchers": [
                [
                  "region",
                  "=",
                  "nasa",
                ],
              ],
              "receiver": "B2",
              "routes": [],
            },
          ],
        },
        {
          "object_matchers": [
            [
              "foo",
              "=",
              "bar",
            ],
          ],
          "receiver": "C",
          "routes": [],
        },
        {
          "continue": undefined,
          "group_by": undefined,
          "group_interval": undefined,
          "group_wait": undefined,
          "match": undefined,
          "match_re": undefined,
          "matchers": undefined,
          "mute_time_intervals": undefined,
          "object_matchers": undefined,
          "receiver": "E",
          "repeat_interval": undefined,
          "routes": undefined,
        },
      ],
    },
    "time_intervals": [],
  },
  "template_files": {},
}
`;

exports[`routes Should delete route 1`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [],
    "route": {
      "group_by": [
        "grafana_folder",
      ],
      "group_interval": "1m",
      "group_wait": "10s",
      "object_matchers": [],
      "receiver": "ROOT",
      "routes": [
        {
          "object_matchers": [
            [
              "foo",
              "=",
              "bar",
            ],
          ],
          "receiver": "C",
          "routes": [],
        },
      ],
    },
    "time_intervals": [],
  },
  "template_files": {},
}
`;

exports[`routes Should update an existing route A with receiver B 1`] = `
{
  "alertmanager_config": {
    "mute_time_intervals": [],
    "route": {
      "group_by": [
        "grafana_folder",
      ],
      "group_interval": "1m",
      "group_wait": "10s",
      "object_matchers": [],
      "receiver": "ROOT",
      "routes": [
        {
          "continue": undefined,
          "group_by": undefined,
          "group_interval": undefined,
          "group_wait": undefined,
          "match": undefined,
          "match_re": undefined,
          "matchers": undefined,
          "mute_time_intervals": undefined,
          "object_matchers": undefined,
          "receiver": "B",
          "repeat_interval": undefined,
          "routes": [
            {
              "object_matchers": [
                [
                  "region",
                  "=",
                  "europe",
                ],
              ],
              "receiver": "B1",
              "routes": [],
            },
            {
              "object_matchers": [
                [
                  "region",
                  "=",
                  "nasa",
                ],
              ],
              "receiver": "B2",
              "routes": [],
            },
          ],
        },
        {
          "object_matchers": [
            [
              "foo",
              "=",
              "bar",
            ],
          ],
          "receiver": "C",
          "routes": [],
        },
      ],
    },
    "time_intervals": [],
  },
  "template_files": {},
}
`;
